src/error.rs

```diff
@@ -42,3 +43,49 @@ pub enum Error {
     #[error("permission denied opening socket {0}")]
     PermissionDeniedOpeningSocket(String),
 }
+
+impl Clone for Error {
+    fn clone(&self) -> Self {
+        match self {
```

Comment:
This all needs to be implemented by hand just because io::Error doesn't implement clone right? What about just dropping io::Error into an Arc and then deriving clone?

Reply:
Basically yes [`io::Error`](https://doc.rust-lang.org/std/io/struct.Error.html) and [`SendError`](https://docs.rs/tokio/latest/tokio/sync/broadcast/error/struct.SendError.html) don't implement `Clone`.

Arc is an interesting idea, I had a play with it this morning, and what I gain from deriving `Clone` I loose from having to manually implement `From` and manually wrapping `UnsolicitedIoError` with `Arc`. The `Arc` version is still less code but it bugs me to add a whole level of indirection and atomic counter wrapped around what it probably just an `i32`.

The reason I added Clone is that I thought it made more sense to deal with error as the response to the client, rather than crashing the main worker task, and so I need to forward all scan errors to all the callers awaiting scan results. My other idea is forward the error to the the first client, then retry getting the scan results so there is no need to clone the error.

I'll add a commit without the cloning and you can see if you think that is nicer.

Reply:
So funny enough I've had the same philosophical debate at work recently about `io::Errors` and fundamentally, there are reasons why they're not clonable already. Within this particular context, I'd boil it down to the following idea: all of these clients are separated by channels from the original `io::Error` and can't do anything with the cloned error. It makes no sense that N client could get the `io::Error` when they can't do anything about it.

You've done a lot of good things in this PR and I'm happy to move along and land it if you make a simple change: create a new Error variant called Error::RemoteIo and your io_clone fn could be
```
fn clone_io(err: &std::io::Error) -> Error {
    Error::RemoteIo(err.to_string())
}
```

Alternatively, if you have the appetite to try something more in depth, I'm not sure the clients calls should return an Error that contains `io::Error`. The `Error` enum might benefit by being separated out so that some Errors apply to the runtimes (WifiAp::run / WifiStation::run) and some apply to the client calls. So this should take a little more time untangling the monolithic error but I think it will have a nice benefit of being clear about what calls actually provide which errors.

Reply:
Actually, please don't do what I suggested in cloning. Perhaps a fair half-measure is doing: "error.for_client()" where you do something like the above. I don't like the idea of having a clone footgun laying around

Reply:
So for now I have side stepped this with [b2eaace](https://github.com/lthiery/wifi-ctrl/pull/15/commits/b2eaace7ec891bd6a10ee71c5b9918831e9b61ae) meaning I don't need to clone errors anymore, as each error goes to exactly one client. 

However I think you have a point around breaking up the giant Error enum and thinking more carefully which errors go where and what the client should do about them.

I glibly decided to forward them back to the client as the errors I was seeing where parsing errors, (I didn't have good enough logs to work out what was failing to be parsed) so the client could deal with the failed request but keep using the connection, however an IO error would imply that something is wrong with the socket so maybe it does make more sense for the runner to fail at that point as the socket probably need reconnecting. That said even for IO errors the client still need some kind of error as otherwise it will hang forever waiting for a response that never comes.

How about I unpick my error handling changes from the UTF8 escaping changes and stick them in a separate PR, then in that PR I have an attempt at creating separate enums for client and runner errors.

Reply:
That sounds like a plan to me!